keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,305 @@
+    * After deprecation period For CSI volumes limits coming from `Node.status.allocatable` and `Node.status.capacity` will be completely ignored.
+    * For in-tree drivers (i.e AWS, Azure, GCE and Cinder) existing predicates will keep working until CSI migration is enabled by default for each driver. Once CSI migration is enabled by default for a volume plugin, after 2 releases the predicate for that volume plugin will be removed from scheduler. For example: If `CSIMigrationGCE` is enabled by default in 1.15 then in 1.17 - scheduler will completly remove `MaxGCEPDVolumeCountPred` predicate.
+    * If `CSINode.spec.drivers[xyz]` (or whole `CSINode`) is missing, scheduler falls back to `Node.status.allocatable` for volume limits. This solves version skew between old kubelet (using `Node.status`) and new scheduler.
+    * The in-tree drivers for which migration has been enabled by default, their in-tree predicate (such as `MaxGCEPDVolumeCountPred`) will be updated to check both `CSINode.Spec.Drivers` and `CSINode` annotation to verify if migration is enabled. If for a in-tree driver migration is enabled but `CSINode.Spec.Drivers` does not have an entry then scheduling decision will be offloaded to `MaxCSIVolumeCountPred` predicate.
+    * After deprecation period, scheduler won't schedule any pods that use volumes to a node with missing `CSINode` instance. It is expected that it happens only during node registration when `Node` exists and `CSINode` doesn't and it self-heals quickly.
+    * After deprecation period, if `CSINode` exists and  does not contain a CSI driver, scheduler will not put CSI volumes for the driver to the node.
+
+CSINode example:
+
+```yaml
+apiVersion: storage.k8s.io/v1beta1
+kind: CSINode
+metadata:
+  name: ip-172-18-4-112.ec2.internal
+spec:
+  drivers:
+    - name: ebs.csi.aws.com
+      # Already existing fields
+      nodeID: ip-172-18-4-112.ec2.internal
+      topologyKeys:
+         # ...
+      # New API:
+      allocatable:
+        # AWS node can attach max. 40 volumes, 1 is reserved for the system
+        volumes: 39
+    - name: org.kernel.nfs
+      allocatable:
+        # NFS does not impose any limits of volumes mounted to the node
+        volumes:    # nil means "no limit"
+```
+
+* Existing scheduler predicates `MaxEBSVolumeCount`, `MaxGCEPDVolumeCount`, `MaxAzureDiskVolumeCount` and `MaxCinderVolumeCount` are already deprecated.
+  * If any of them is enabled together with `MaxCSIVolumeCountPred`, the deprecated predicate will do nothing (`MaxCSIVolumeCountPred` does the job of counting both in-tree and CSI volumes).
+  * The deprecated predicates will do any useful work only when `MaxCSIVolumeCountPred` predicate is disabled or `CSINode` object does not have limits for a particular driver.
+  * This way, we save CPU by running only one volume limit predicate during deprecation period.
```

Comment:
How do the deprecated predicates know to exit early?  I assume they will have to do some work to translate and check the CSINode object

Reply:
We don't need to look at actual volumes though. Since existing predicates are volume-type specific we can just look at `CSINode` object and translate the name and return early from the predicate.

Reply:
strictly speaking the old predicate is not a noop, it still has to do a lookup, although it is much better than having to iterate over all pods

Reply:
Okay I have added a note for what `NOOP` means in this context